### PR TITLE
Fix Full Framework Tests issues when running in Docker container

### LINKED ISSUE #200: Tests for Full Framework code
We agreed that it's would okay to merge https://github.com/elastic/apm-agent-dotnet/pull/195 PR 
without tests covering main added functionality and we will do it in a separate PR.

### PR DESCRIPTION
**Pull request by SergeyKleyman**
Closes #200

Continuation of #284

The focus of this PR is to fix Full Framework Tests issues when running in Docker container. While investigating those issues I had to add a few debuggability features:
- Added NLog support `AspNetFullFrameworkSampleApp` for application's and agent's logging. It required a very small addition to public API to allow setting logging "backend" for agent to use. It's very far from ideal and I opened elastic/apm-agent-dotnet#318 to follow up.
- Sample application failed to write logs to a file because of file system permissions but in order to see the logs in some way I added /Diagnostics page to `AspNetFullFrameworkSampleApp` which shows logs accumulated in memory by NLog's [`MemoryTarget`](https://github.com/NLog/NLog/wiki/Memory-target)
- After finally getting to agent's logs it turned out that application used wrong values in environment variable in which we pass configuration settings. In particular wrong value in `ELASTIC_APM_SERVER_URLS` prevent agent from connecting to mock APM server and sending any data. In addition to implementing the fix I've added a mechanism to detect without any investigation it if it ever happens again. I've added a mechanism for `AspNetFullFrameworkSampleApp` to add `ELASTIC_APM_SERVER_URLS` value as response header so unit tests can compare it against expected value.

@kuisathaverat Please take a look and let's discuss how to finalize "Full Framework Tests in Docker container" I did merge elastic/apm-agent-dotnet#284 into this one but maybe you wanted to add more CI infrastructure related code that is why I marked this PR as WiP for now. Please note that from all the changes the one of interest to you is in `test-iis.bat` - to address an issue I've mentioned above when file system permissions  did allow sample application running under IIS's restricted user account to write to log file.

### CODE REVIEW

**Comment by gregkalapos** [comment]
Don't wanna interfere here, but this seems getting giant with a lots of things going on - any plan to get to a reviewable state soon?  - just to make sure we merge things that are fairly obvious without being blocked by unrelated things.

**Comment by SergeyKleyman** [comment]
All these commits are just to keep this PR in sync with master but I don't add anything new to it. 

It's in review-able state from agent's code point of view - the reason I'm keeping it marked as draft is because infrastructure is still not ready so there is some slim chance some changes to agent's code will be required although it's very unlikely because I run this PR in a docker container on my laptop in the background all the time.

So if you think it's preferable to start review now before CI infrastructure is finished I am ready to start.

**Comment by SergeyKleyman** [comment]
We can review and merge only agent's code (excluding changes to CI infrastructure that I merged from #284). The tests will be in master but skipped. When infrastructure teams is ready they will merge their change enabling the tests. WDYT?

**Comment by kuisathaverat** [comment]
Noted https://github.com/elastic/apm-agent-dotnet/pull/284 was superseded by https://github.com/elastic/apm-agent-dotnet/pull/415, in that PR we replaced the initial approach to run the tests on Docker for the approach of running the tests on an ephemeral worker. If you do not need any change on the https://github.com/elastic/apm-agent-dotnet/blob/master/.ci/windows/test-iis.bat script to run the test, the current changes on the Jenkinsfile would be enough to launch the test.

File: sample/AspNetFullFrameworkSampleApp/Global.asax.cs [hunk 1]
```diff
@@ -1,21 +1,50 @@
...
-            BundleConfig.RegisterBundles(BundleTable.Bundles);
-        }
-    }
+	public class MvcApplication : HttpApplication
+	{
+		protected void Application_Start()
+		{
+			LoggingConfig.SetupLogging();
+
+			var Logger = LogManager.GetCurrentClassLogger();
```
[the remark this answers is not in the document]
Comment (SergeyKleyman) [reply sample/AspNetFullFrameworkSampleApp/Global.asax.cs]: Done.

File: sample/AspNetFullFrameworkSampleApp/App_Start/ApmLoggerToNLog.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace AspNetFullFrameworkSampleApp
+{
+	internal class ApmLoggerToNLog : IApmLogger
+	{
+		public bool IsEnabled(ApmLogLevel level) => true;
+
+		public void Log<TState>(ApmLogLevel apmLogLevel, TState state, Exception e, Func<TState, Exception, string> formatter)
+		{
+//			var stateAsKeyValuePairs = state as ReadOnlyCollection<KeyValuePair<string, object>> ??
```
[the remark this answers is not in the document]
Comment (SergeyKleyman) [reply sample/AspNetFullFrameworkSampleApp/App_Start/ApmLoggerToNLog.cs]: Done.

File: src/Elastic.Apm/Agent.cs [hunk 3]
```diff
@@ -81,14 +81,32 @@ public void Dispose()
...
+				if (Lazy.IsValueCreated) throw new Exception("The singleton APM agent has already been instantiated and can no longer be configured");
+
+				_logger = value;
+			}
+		}
+
+		public static AgentComponents LastSetupComponents { get; private set; }
+
+		private static AgentComponents BuildComponents()
+		{
```
[the remark this answers is not in the document]
Comment (SergeyKleyman) [reply src/Elastic.Apm/Agent.cs]: Done.

File: src/Elastic.Apm/Report/PayloadSenderV2.cs [hunk 4]
```diff
@@ -54,6 +54,7 @@ public PayloadSenderV2(IApmLogger logger, IConfigurationReader configurationRead
 			servicePoint.ConnectionLeaseTimeout = DnsTimeout;
 			servicePoint.ConnectionLimit = 20;
 
+			_logger?.Debug()?.Log("Setting HTTP client BaseAddress to {ApmServerUrl}...", serverUrlBase, Process.GetCurrentProcess().Id);
```
[the remark this answers is not in the document]
Comment (SergeyKleyman) [reply src/Elastic.Apm/Report/PayloadSenderV2.cs]: Fixed.

File: src/Elastic.Apm/Agent.cs [hunk 5]
```diff
@@ -81,14 +81,32 @@ public void Dispose()
...
+		private static readonly Lazy<ApmAgent> Lazy = new Lazy<ApmAgent>(() => new ApmAgent(BuildComponents()));
+		private static IApmLogger _logger;
 
 		public static IConfigurationReader Config => Lazy.Value.ConfigurationReader;
 
 		internal static ApmAgent Instance => Lazy.Value;
 
 		internal static bool IsInstanceCreated => Lazy.IsValueCreated;
+		
+		public static IApmLogger Logger
```
[the remark this answers is not in the document]
Comment (gregkalapos) [reply src/Elastic.Apm/Agent.cs]: Just to document our discussion:

We agreed to make sure we don't change the API here. We have some alternative approaches  to address this. One was to use `HttpApplication.Context.Items` like [here](https://stackoverflow.com/questions/1657473/ioc-dependency-injection-into-custom-http-module-how-asp-net), and Sergey also had another idea. I think both are fine, once we make sure we don't change the public API for this I think this is ready to merge.

**Comment by gregkalapos** [comment]
Should be merged after #487, since this contains changed from that PR.

**Pull request merged**